src/main/java/com/squareup/javawriter/builders/TypeSpec.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.javawriter.builders;
+
+import com.google.common.collect.ImmutableList;
+import com.squareup.javawriter.ClassName;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/** A generated class, interface or enum declaration. */
+public final class TypeSpec {
+  public final Type type;
+  public final ClassName name;
+  public final ImmutableList<MethodSpec> methodSpecs;
+
+  private TypeSpec(Builder builder) {
+    this.type = checkNotNull(builder.type);
+    this.name = checkNotNull(builder.name);
+    this.methodSpecs = ImmutableList.copyOf(builder.methodSpecs);
+  }
+
+  void emit(CodeWriter codeWriter) {
+    codeWriter.emit("class $L {\n", name.simpleName()); // TODO: modifiers.
+    codeWriter.indent();
+
+    boolean firstMethod = true;
+    for (MethodSpec methodSpec : methodSpecs) {
+      if (!firstMethod) codeWriter.emit("\n");
+      methodSpec.emit(codeWriter);
+      firstMethod = false;
+    }
+
+    codeWriter.unindent();
+    codeWriter.emit("}\n");
+  }
+
+  public static enum Type {
```

Comment:
While I agree and love the type safety afforded by the current JW3 code, we also have javac at our backs. We will never be able to prevent 100% of invalid code in the type system so it becomes a balance of how much we put into the API and how much we just defer to javac.

Ultimately I think we should absolutely explore subtypes of this class to map to the different Java types.